run-nextflow-tests/configtest.py

```diff
@@ -255,27 +258,67 @@ def print_diffs(self, other: T):
 
             print(f"::{level} {annotation}::{diff}")
 
-    def mark_for_archive(self):
+    def mark_for_archive(self, test_passed: bool, prior: T):
         "Emit GitHub workflow commands to archive this file."
-        if "GITHUB_OUTPUT" not in os.environ:
-            # Nothing to be done here
-            print("Unable to mark for archive")
-            return
+        relpath = str(self.filepath.relative_to(self.pipeline))
+        # Each archive file needs a unique key
+        key = BAD_CHARACTERS.sub("_", relpath)
+
+        # https://github.com/actions/upload-artifact#upload-using-multiple-paths-and-exclusions
+        # https://github.com/actions/upload-artifact/issues/174#issuecomment-1909478119
+        # Sigh. We also need to include a dummy file at the root to ensure
+        # that the directory structure is preserved
+        dummy_filename = f".{key}-dummy"
+        Path(dummy_filename).touch()
+
+        filenames = [relpath, dummy_filename]
+
+        if not test_passed:
+            # Also save out a file at the repository root with the `jd` output
+            with tempfile.TemporaryDirectory() as tempdir:
+                before = Path(tempdir, "before.json")
+                after = Path(tempdir, "after.json")
+
+                with before.open(mode="w", encoding="utf-8") as outfile:
+                    json.dump(prior.expected_result, outfile)
+
+                with after.open(mode="w", encoding="utf-8") as outfile:
+                    json.dump(self.expected_result, outfile)
```

Comment:
Where does `self.expected_result` actually get set?

I'm having a little trouble following the exact flow here.

I see that we have these two concepts of "a prior expected result" and some "expected result" that we write to "before" and "after" (respectively). The prior result comes from wherever called this method, but the other expected result is a class member and set "globally"?

Then we run `jd` and add a collection of EOF blocks to a file.

Perhaps the class-member `expected_result` cannot be changed in this PR (out of scope). I think some minor changes to variable names in this method, and perhaps splitting this method into a couple different methods might clarify what's happening. Some possible candidates:

- `outfile` is redefined four times in this method; without scanning the entire method to see when it changes, it's easy to hold the wrong file in mind when reading the code.
- What are "before" and "after"? I think, knowing what this PR is about, we're looking at some PR content "before" any automation ran, and then PR content "after"?
- Why does this only run if `not test_passed`? Could this be presented as something like `write_pr_failure_notes` or some such? I think this method doesn't "know" what passing or failing tests are, but the caller of it does, so allowing the caller to decide whether to "write PR notes" based on its knowledge of failing tests may clarify the logic of whether to run this block.
- Because there are two sort of "modes" here - writing more data if `not test_passed`, and writing data either way, splitting the `not test_passed` block into a separate method may simplify reading this method.

Lastly, is there somewhere obvious I missed that explains what the EOF blocks are for? Maybe a short comment directing to information on what they are can help.

Reply:
Firstly, the `EOF` nonsense is used to [pass multiline-outputs](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#multiline-strings) to `GITHUB_OUTPUT`. The `eof_index` chicanery is to make sure that filenames with `EOF` in them don't cause problems.

Reply:
At a high level, here is the general flow:

1. A `NextflowConfigTest` (a dataclass) is [parsed from a file](https://github.com/uclahs-cds/tool-Nextflow-action/blob/43877d5401e3240e80863a3d4d8bd400a8c0f86f/run-nextflow-tests/entry.py#L13)
2. `recompute_results()` [generates a new dataclass instance](https://github.com/uclahs-cds/tool-Nextflow-action/blob/43877d5401e3240e80863a3d4d8bd400a8c0f86f/run-nextflow-tests/entry.py#L18) by running the tests. [Internally](https://github.com/uclahs-cds/tool-Nextflow-action/blob/43877d5401e3240e80863a3d4d8bd400a8c0f86f/run-nextflow-tests/configtest.py#L323-L339) that does a [little bit more](https://github.com/uclahs-cds/tool-Nextflow-action/blob/43877d5401e3240e80863a3d4d8bd400a8c0f86f/run-nextflow-tests/configtest.py#L63-L78) than `dataclasses.replace()`. _This is where `expected_results` gets updated_.

From here it gets a little more complicated, I'll handle that in a separate comment.

Reply:
I reworked a lot of this with 0bcfd6a (and 43f7448) with a more streamlined flow and more comments. Does it look better to you now?

Reply:
Ahh, and [that's](https://github.com/uclahs-cds/tool-Nextflow-action/actions/runs/8525747220/job/23353433322) why I re-used `outfile` so much - that method is right on the edge of pylint's [too-many-locals](https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/too-many-locals.html) limit.

Reply:
> ... that method is right on the edge of pylint's [too-many-locals](https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/too-many-locals.html) limit.

Ah, interesting! I think this is a good example of a linter's limitations and when their "objective" nature starts to tread into subjectiveness. Short of endless heuristics, how can it decide whether it's better to reuse a local to stay under that limit, or create a new one to increase understanding? How can it prescribe what "too many locals" is for _every_ method? I have my own opinions about what's "best" as I'm sure we all do. ¯\\\_(ツ)_/¯

Reply:
> I reworked a lot of this with [0bcfd6a](https://github.com/uclahs-cds/tool-Nextflow-action/commit/0bcfd6ae562fbf78a46269888af055d91ab667dc) (and [43f7448](https://github.com/uclahs-cds/tool-Nextflow-action/commit/43f7448487315835dce31ae624df2310c44bc938)) with a more streamlined flow and more comments. Does it look better to you now?

Yeah, looks great! It's quite clear what's happening, and both the `print_only` variable name and its explanation bring in context that I wasn't previously aware of, so that's awesome.

I just have one comment I added.